commands/build_bundles.py

```diff
@@ -0,0 +1,202 @@
+
+
+@retry_timeout
+def get_modified_timestamp(url):
+    """
+    Return URL modified date as epoch millisecond.
+    """
+    resp = requests.get(url)
+    if not resp.ok:
+        return None
+    dts = resp.headers["Last-Modified"]
+    dt = parsedate_to_datetime(dts)
+    epoch_msec = int(dt.timestamp() * 1000)
+    return epoch_msec
+
+
+@retry_timeout
+def fetch_attachment(url):
+    print("Fetch %r" % url)
+    resp = requests.get(url)
+    return resp.content
+
+
+def write_zip(output_path: str, content: list[tuple[str, bytes]]):
+    """
+    Write a Zip at the specified `output_path` location with the specified `content`.
+    The content is specified as a list of file names and their binary content.
+    """
+    parent_folder = os.path.dirname(output_path)
+    os.makedirs(parent_folder, exist_ok=True)
+
+    zip_buffer = io.BytesIO()
+    with zipfile.ZipFile(zip_buffer, "w", zipfile.ZIP_DEFLATED) as zip_file:
+        for filename, filecontent in content:
+            zip_file.writestr(filename, filecontent)
+    with open(output_path, "wb") as f:
+        f.write(zip_buffer.getvalue())
+    print("Wrote %r" % output_path)
+
+
+def sync_cloud_storage(
+    storage_bucket: str, remote_folder: str, to_upload: list[str], to_delete: list[str]
+):
+    """
+    Upload the specified `to_upload` filenames, and delete the specified `to_delete` filenames
+    from the `remote_folder` of the `storage_bucket`.
+    """
+    # Ensure you have set the GOOGLE_APPLICATION_CREDENTIALS environment variable
+    # to the path of your Google Cloud service account key file before running this script.
+    client = storage.Client()
+    bucket = client.bucket(storage_bucket)
+    for filename in to_upload:
+        remote_file_path = os.path.join(remote_folder, filename)
+        blob = bucket.blob(remote_file_path)
+        blob.upload_from_filename(filename)
+        print(f"Uploaded {filename} to gs://{storage_bucket}/{remote_file_path}")
+
+    to_delete = {os.path.join(remote_folder, f) for f in to_delete}
+    blobs = bucket.list_blobs(prefix=remote_folder)
+    for blob in blobs:
+        if blob.name in to_delete:
+            blob.delete()
+            print(f"Deleted gs://{storage_bucket}/{blob.name}")
+
+
+def build_bundles(event, context):
+    """
+    Main command entry point that:
+    - fetches all collections changesets
+    - builds a `changesets.zip`
+    - fetches attachments of all collections with bundle flag
+    - builds `{bid}--{cid}.zip` for each of them
+    - send the bundles to the Cloud storage bucket
+    """
+    rs_server = event.get("server") or SERVER
+
+    client = KintoClient(server_url=rs_server)
+
+    base_url = client.server_info()["capabilities"]["attachments"]["base_url"]
+
+    existing_bundle_timestamp = get_modified_timestamp(
+        f"{base_url}{DESTINATION_FOLDER}/changesets.zip"
+    )
+    if existing_bundle_timestamp is None:
+        print("No previous bundle found")  # Should only happen once.
```

Comment:
This could happen multiple times.Ex: if the bundle cannot be built due it being too large.

Reply:
The `changesets.zip` will always be produced (and it's currently ~6MB) so once the job has ran once, the file will be there